assembly/src/parsers/ast/mod.rs

```diff
@@ -0,0 +1,593 @@
+        tokens.advance();
+
+        let mut loop_body = Vec::<Node>::new();
+        // read the loop body
+        self.parse_body(tokens, &mut loop_body)?;
+
+        // consume the `end` token
+        match tokens.read() {
+            None => Err(AssemblyError::unmatched_repeat(
+                tokens.read_at(repeat_start).expect("no repeat token"),
+            )),
+            Some(token) => match token.parts()[0] {
+                Token::END => token.validate_end(),
+                Token::ELSE => Err(AssemblyError::dangling_else(token)),
+                _ => Err(AssemblyError::unmatched_repeat(
+                    tokens.read_at(repeat_start).expect("no repeat token"),
+                )),
+            },
+        }?;
+        tokens.advance();
+
+        Ok(Node::Repeat(repeat_start, loop_body))
+    }
+
+    // parse exec token into AST nodes.
+    fn parse_exec(&self, label: String, tokens: &mut TokenStream) -> Result<Node, AssemblyError> {
+        tokens.advance();
+
+        if label.contains(MODULE_PATH_DELIM) {
+            let full_proc_name = self.get_full_imported_proc_name(label);
+            Ok(Node::Instruction(Instruction::ExecImported(full_proc_name)))
+        } else {
+            let index = self.procedures.get(&label).unwrap().index;
+            Ok(Node::Instruction(Instruction::ExecLocal(index)))
+        }
+    }
+
+    // parse call token into AST nodes.
+    fn parse_call(&self, label: String, tokens: &mut TokenStream) -> Result<Node, AssemblyError> {
+        tokens.advance();
+        if label.contains(MODULE_PATH_DELIM) {
+            let full_proc_name = self.get_full_imported_proc_name(label);
+            Ok(Node::Instruction(Instruction::CallImported(full_proc_name)))
+        } else {
+            let index = self.procedures.get(&label).unwrap().index;
+            Ok(Node::Instruction(Instruction::CallLocal(index)))
+        }
+    }
+
+    fn get_full_imported_proc_name(&self, short_name: String) -> String {
+        let (module_name, proc_name) = short_name.split_once(MODULE_PATH_DELIM).unwrap();
+        let full_module_name = self.imports.get(module_name).unwrap();
+        format!("{}{}{}", full_module_name, MODULE_PATH_DELIM, proc_name)
+    }
+
+    // Parse procedures in the source and store them in the program
+    fn parse_procedures(
+        &mut self,
+        tokens: &mut TokenStream,
+        allow_export: bool,
+    ) -> Result<(), AssemblyError> {
+        while let Some(token) = tokens.read() {
+            match token.parts()[0] {
+                Token::EXPORT => {
+                    if !allow_export {
+                        let (label, _, _) = token.parse_proc()?;
+                        return Err(AssemblyError::proc_export_not_allowed(token, &label));
+                    }
+                    self.parse_procedure(tokens)?;
+                }
+                Token::PROC => {
+                    self.parse_procedure(tokens)?;
+                }
```

Comment:
nit: I'd probably change this to something like:
```Rust
Token::PROC | Token::EXPORT => {
    let proc = self.parse_procedure(tokens)?;
    if !allow_exports && proc.is_export {
        return Err(AssemblyError::proc_export_not_allowed(token, &proc.name));        
    }
    self.procedures.insert(proc.name.to_string(), proc);
}
```
Also, we should probably check here that procedures don't have duplicate names.

Reply:
fixed